File: apps/web/src/app/(app)/sell/listing-text.ts
import type { Item } from "@loop/contracts";
import { formatCents } from "@/lib/format";

type ValuationRow = { estimated_value_cents: number };
type ListingRow = { marketplace: string; status: string; list_price_cents: number | null };

/**
 * The one canonical text representation of an item's listing, shared by
 * Copy/Share/Export. Mirrors `buildListingText` in
 * apps/mobile/lib/features/sell/listing_text.dart field-for-field.
 *
 * Deliberately excludes ids, account ids, Storage paths, and every other
 * backend/authorization detail — this text is meant to leave the device.
 */
export function buildListingText(
  item: Pick<Item, "name" | "category" | "condition">,
  valuation: ValuationRow | undefined,
  listings: ListingRow[],
): string {
  const lines = [item.name];

  const details = [item.category, item.condition].filter(Boolean).join(" · ");
  if (details) lines.push(details);

  const activeListing = listings.find((l) => l.status === "draft" || l.status === "active");
  if (activeListing) {
    lines.push(
      activeListing.list_price_cents != null
        ? `Asking ${formatCents(activeListing.list_price_cents)} on ${activeListing.marketplace}`
        : `Listed on ${activeListing.marketplace}`,
    );
  } else if (valuation) {
    lines.push(`Estimated value ${formatCents(valuation.estimated_value_cents)}`);
  }

  return lines.join("\n");
}

File: apps/mobile/lib/features/sell/listing_text.dart
import '../../core/utils/money.dart';
import 'models/item.dart';
import 'models/listing.dart';
import 'models/valuation.dart';

/// The one canonical text representation of an item's listing, shared by
/// Copy/Share/Export. Mirrors `buildListingText` in
/// apps/web/src/app/(app)/sell/listing-text.ts field-for-field.
///
/// Deliberately excludes ids, account ids, Storage paths, and every other
/// backend/authorization detail — this text is meant to leave the device.
String buildListingText({
  required Item item,
  ValuationRow? valuation,
  List<ListingRow> listings = const [],
}) {
  final lines = <String>[item.name];

  final details = [
    item.category,
    item.condition,
  ].whereType<String>().where((s) => s.isNotEmpty).join(' · ');
  if (details.isNotEmpty) lines.add(details);

  final activeListing = listings.where(
    (l) => l.status == ListingStatus.draft || l.status == ListingStatus.active,
  );
  if (activeListing.isNotEmpty) {
    final listing = activeListing.first;
    final price = listing.listPriceCents;
    lines.add(
      price != null
          ? 'Asking ${MoneyUtils.formatCents(price)} on ${listing.marketplace}'
          : 'Listed on ${listing.marketplace}',
    );
  } else if (valuation != null) {
    lines.add(
      'Estimated value ${MoneyUtils.formatCents(valuation.estimatedValueCents)}',
    );
  }

  return lines.join('\n');
}
